[WIP] MM-11327: Restrict Teams by Email

gvengel opened the pull request:
#### Summary
This allows individual teams to be restricted to an email domain whitelist. The per team setting overrides the server-wide whitelist, but teams which do not set the option will still fallback to the server-wide setting.

The change leverages the preexisting team.allowedDomains model property. For now, it must be modified via REST API, as UI has not been implemented.

I started modifying the SanitizeTeam code to permit updates to allowedDomains, but wasn't sure if that should be included in this PR or later when UI is added. I left it out for now, but can update the PR, if appropriate?

Tests have been added for the AddUserToTeam* methods to make sure bad email domains get blocked when the whitelist is enabled. I'm unsure if there are other code paths which may need testing. Open vs invite only teams, or enterprise LDAP accounts, for example.

#### Ticket Link
- Help Wanted: #9111
- Uservoice: [restrict-team-by-email](https://mattermost.uservoice.com/forums/306457-general/suggestions/33014368-restrict-team-by-email)

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)

gvengel commented:
Ah, yes, looks like I broke the sanitization tests under api4. I'll get that sorted out.

gvengel commented:
Okay, I've gone ahead and removed AllowedDomains from the team sanitization code. This both resolves the tests, and preps the code so AllowedDomains can be updated via the UI.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-06767d016a0b22aab.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-06767d016a0b22aab

wiersgallak commented:
Thanks again for your work on this @gvengel!  We will have Jesse review from a technical point of view.  From the product perspective, we'd like to have this available in the UI as well prior to implementation.  I will coordinate some design resources to provide more guidance on the UI.

gvengel commented:
Thanks @wiersgallak. From a technical point of view, we had talked in the pre-release dev channel about shipping this as a REST only beta. I understand if you would rather wait until the UI has been implemented, and I'm happy to try and figure it out. I'm new to React, so it might be slow going. Any help would be appreciated. I'm hoping I can just copypasta some code in the team edit UI. 😄

gvengel commented:
Sweet, I was able to get the UI working by copying existing code. Once I have all the kinks worked out, how do I submit the changes? Do I create a separate PR against mattermost-webapp and link it to this PR?

wiersgallak commented:
@gvengel That is correct and thank you taking the time to add the UI. I have this queued for our designers to mock up a design, but if you have the UI work completed, I can have them provide feedback to you on the PR directly.

lieut-data commented on app/team.go:
```diff
@@ -430,6 +434,9 @@ func (a *App) joinUserToTeam(team *model.Team, user *model.User) (*model.TeamMem
 
 func (a *App) JoinUserToTeam(team *model.Team, user *model.User, userRequestorId string) *model.AppError {
+	if !a.isTeamEmailAllowed(user, team) {
+		return model.NewAppError("isTeamEmailAllowed", "api.team.is_team_creation_allowed.domain.app_error", nil, "", http.StatusBadRequest)
```
The convention here is to name the function in which it failed, e.g. passing `JoinUserToTeam` as the first parameter.

gvengel replied:
Fixed.

lieut-data commented on app/team_test.go:
```diff
@@ -96,13 +96,33 @@ func TestAddUserToTeam(t *testing.T) {
 
-	user := model.User{Email: strings.ToLower(model.NewId()) + "success+test@example.com", Nickname: "Darth Vader", Username: "vader" + model.NewId(), Password: "passwd1", AuthService: ""}
-	ruser, _ := th.App.CreateUser(&user)
+	t.Run("add user", func(t *testing.T) {
```
Can you add some coverage for
* `th.BasicTeam.AllowedDomains` is non-empty but the user domain matches
* `th.BasicTeam.AllowedDomains` has multiple domains
* `th.BasicTeam.AllowedDomains` is non-empty and the user has a subdomain of a valid domain? (Should reject it.)
* Tests that include setting `RestrictCreationToDomains` and the interplay between both (e.g. team override, but system disallows).

Also, it looks like there's some coverage in `api4/team_test.go` that tests `RestrictCreationToDomains`. It's unfortunate the way our code is split up, but I suspect it makes the most sense to co-locate these tests together. 0/5 on whether we extend the tests here or add to the tests in `api4/team_test.go`.

gvengel commented:
@lieut-data Re: team vs global setting. Ah yes, good catch. My original version disabled the global setting if the team was enabled, but I completely forgot when I was writing up the PR notes that I'd later changed my mind and applied both. My thinking was it would be unexpected to a system admin if a team admin could override the global setting and open the server up to more users without the system admin's approval.

Additionally, since the global setting isn't exposed to a team admin, they can currently set a team to foo.com, when the global setting only allows bar.com. In this case *no* email would be valid. Trying to enforce consistency between the two settings seems tricky. Instead maybe the team edit UI should just throw a warning or error if the team setting is saved with a domain which is not in the global setting?

gvengel commented:
@lieut-data Re: sanitizing AllowedDomains. To be honest, I don't fully understand the sanitization process. My assumption was that sanitization was used to prevent form updates to properties which were considered read only. I'll dig into the code and try to understand it better, but any pointers would be helpful.

gvengel commented:
@lieut-data Okay, I think I've addressed all your change requests. You were right about sanitizing AllowedDomains, so I backed that change out. Let me know if there is anything else which needs attention. Thanks!

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

gvengel commented:
Rebased for @wiersgallak.

mattermod commented:
Spinmint test server destroyed

gvengel commented:
Hmm, there seems to be a bug when changing the AllowedDomains property via UI that causes it to randomly clear the setting when saving some times. I've tested this on two separate systems with a clean build.

@lieut-data @wiersgallak Can you reproduce the issue on your end?

gvengel commented:
I ran some manual tests with the API, and didn't see this behavior. There is something non-deterministic in the UI update causing it to clear the setting. I'm guessing it may be related to sanitization?

gvengel commented:
Ah, more data. It looks like the failure is cosmetic only. The UI is refreshed with an empty value, but if I query the team via API the change has actually persisted property. I'll dig around in team_general_tab.jsx and see if I can find a root cause.

gvengel commented:
I did some digging over the weekend, and still couldn't isolate this issue. Any help would be appreciated. @jasonblais @lieut-data @wiersgallak

lieut-data commented:
@gvengel, @wiersgallak: absolutely -- I'll take a look :)

gvengel commented:
Thanks @lieut-data, happy hunting! 👍

lieut-data commented:
Ok, I've tracked this one down, and unfortunately it is related to the sanitization we discussed earlier. It turns out our websocket event handling only distinguishes system admins from non-system admins, and doesn't currently have the wherewithal to avoid sanitizing this for affected team admins.

I'm following up internally on this one to figure out the best course of action, and will post back ASAP with a proposed plan :)

gvengel commented:
Cool, thanks for investigating. I was thinking it was something along those lines, but couldn't figure out why it didn't fail consistently. It's worth noting, I was using a system admin account in my tests, and was still seeing the occasional failure.

lieut-data commented:
@gvengel, just wanted to assure you that this PR has not been lost :) We were out a few people last week from whom I wanted an opinion on this issue, but I'm hopeful we'll make progress on this front shortly.

wiersgallak commented:
@gvengel Wanted to give you an update that this is queued for discussion this week.  We will get back to you as soon as we have some additional guidance to provide.

hmhealey commented:
@gvengel @lieut-data For the sanitization issue, we've decided that it's fine to send the AllowedDomains to all users who have the team. We don't consider it a security concern if someone knows that you need to have a specific email domain to join a given team.

@gvengel If you wouldn't mind removing that line from `Team.Sanitize`, we'll be good to go with that. I'm just going to do a review right now since that shouldn't change much

hmhealey commented on app/team.go:
```diff
@@ -430,6 +455,9 @@ func (a *App) joinUserToTeam(team *model.Team, user *model.User) (*model.TeamMem
 
 func (a *App) JoinUserToTeam(team *model.Team, user *model.User, userRequestorId string) *model.AppError {
+	if !a.isTeamEmailAllowed(user, team) {
+		return model.NewAppError("JoinUserToTeam", "api.team.is_team_creation_allowed.domain.app_error", nil, "", http.StatusBadRequest)
```
This should have its own error message in `i18n/en.json`

gvengel commented:
@hmhealey Sounds good! I've gone ahead and removed AlloweDomains from sanitization, added the i18n string, and rebased on master. Let me know if you need anything else.

wiersgallak commented:
@cpanato can you restart this spinmint please? @gvengel since this has been out there for a bit, I'd like to take another look at this one :)

gvengel commented:
Sounds good, just let me know if you need anything from my end.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0c89b498a098aeda4.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0c89b498a098aeda4

cpanato commented:
@wiersgallak the test server is running, not need a restart

mattermod commented:
Spinmint test server destroyed

hmhealey commented:
Merging as the PM testing will be done on the web app PR
